File: snippets/node/90-lab/06-share-link.mjs
// import the shared client:  import { connect } from "elo-playground";
// topic:    Create a public share link for a document
// category: Testing lab
// id:       lab.share-link

import { connect, EloError } from "elo-playground";

// --- local ELO test box (override with ELOPG_* env vars or a .env) ---
const ELO_BASE_URL = process.env.ELOPG_ELO_BASE_URL || "http://localhost:9090/ix-Repository1"; // ELOPG_DEFAULT:base_url
const ELO_USER = process.env.ELOPG_ELO_USER || "Administrator"; // ELOPG_DEFAULT:user
const ELO_PASS = process.env.ELOPG_ELO_PASSWORD || "elo"; // ELOPG_DEFAULT:password

const elo = await connect({ baseUrl: ELO_BASE_URL, user: ELO_USER, password: ELO_PASS });
const ALL = "449304431574384639";

// --- provision: upload a throwaway document ---
const body = Buffer.from("public share demo\n", "utf-8");
const sord = (await elo.call("createDoc", {
  parentId: 1, maskId: 0, editInfoZ: { bset: "1", sordZ: { bset: ALL } },
})).sord;
sord.name = "pg-share.txt";
const doc = await elo.call("checkinDocBegin", { sord, document: { docs: [{ ext: "txt" }] } });
doc.docs[0].uploadResult = await elo.upload(doc.docs[0].url, body);
const objId = String((await elo.call("checkinDocEnd", {
  sord, document: doc, sordZ: { bset: ALL }, unlockZ: { bset: "1" },
})).objId);
console.log("uploaded doc", objId);

try {
  // 1) mint a public download link
  const pd = await elo.call("insertPublicDownload", {
    opts: { objId, remaining: 5, fileNameFromSordName: true },
  });
  console.log("share url :", pd.url);
  console.log("remaining :", pd.remaining);

  // 2) list the links for this object
  const links = await elo.call("getPublicDownloads", { opts: { objId } });
  console.log("links now :", links.length);

  // 3) revoke them
  await elo.call("terminatePublicDownloadUrls", { opts: { objId } });
  console.log("after revoke:", (await elo.call("getPublicDownloads", { opts: { objId } })).length);
} catch (exc) {
  if (exc instanceof EloError) console.log("share-link operation failed:", exc.message);
  else throw exc;
} finally {
  for (const step of [false, true]) {
    try {
      await elo.call("deleteSord", { objId, parentId: "1", deleteOptions: { deleteFinally: step } });
    } catch (e) { /* best effort */ }
  }
  console.log("cleaned up");
  elo.close();
}

File: snippets/python/90-lab/06-share-link.py
# import the shared client:  from elo_playground import connect
# topic:    Create a public share link for a document
# category: Testing lab
# id:       lab.share-link

import os
from elo_playground import connect, EloError

# --- local ELO test box (override with ELOPG_* env vars or a .env) ---
ELO_BASE_URL = os.getenv("ELOPG_ELO_BASE_URL", "http://localhost:9090/ix-Repository1")
ELO_USER = os.getenv("ELOPG_ELO_USER", "Administrator")
ELO_PASS = os.getenv("ELOPG_ELO_PASSWORD", "")

elo = connect(base_url=ELO_BASE_URL, user=ELO_USER, password=ELO_PASS)
ALL = "449304431574384639"

# --- provision: upload a throwaway document ---
body = b"public share demo\n"
sord = elo.call("createDoc", {"parentId": 1, "maskId": 0,
                              "editInfoZ": {"bset": "1", "sordZ": {"bset": ALL}}})["sord"]
sord["name"] = "pg-share.txt"
doc = elo.call("checkinDocBegin", {"sord": sord, "document": {"docs": [{"ext": "txt"}]}})
doc["docs"][0]["uploadResult"] = elo.upload(doc["docs"][0]["url"], body)
obj_id = str(elo.call("checkinDocEnd", {"sord": sord, "document": doc,
                                        "sordZ": {"bset": ALL}, "unlockZ": {"bset": "1"}})["objId"])
print("uploaded doc", obj_id)

try:
    # 1) mint a public download link (5 downloads, no expiry)
    pd = elo.call("insertPublicDownload", {"opts": {
        "objId": obj_id, "remaining": 5, "fileNameFromSordName": True,
    }})
    print("share url :", pd.get("url"))
    print("remaining :", pd.get("remaining"))

    # 2) list the links for this object
    links = elo.call("getPublicDownloads", {"opts": {"objId": obj_id}})
    print("links now :", len(links))

    # 3) revoke them
    elo.call("terminatePublicDownloadUrls", {"opts": {"objId": obj_id}})
    print("after revoke:", len(elo.call("getPublicDownloads", {"opts": {"objId": obj_id}})))
except EloError as exc:
    print("share-link operation failed:", exc)
finally:
    for step in (False, True):
        try:
            elo.call("deleteSord", {"objId": obj_id, "parentId": "1",
                                    "deleteOptions": {"deleteFinally": step}})
        except EloError:
            pass
    print("cleaned up")
    elo.close()
